Cleanup error handlers

nnethercote opened the pull request:
Mostly by making function naming more consistent. More to do after this, but this is enough for one PR.

r? compiler-errors

nnethercote committed:
Inline and remove `DiagnosticBuilder::new_diagnostic_fatal`.

It has a single call site.

nnethercote committed:
Return `ErrorGuaranteed` from `span_err_with_code` methods.

`ErrorGuaranteed` should be used for all error methods involving the
`Error` level, e.g. as is done for the corresponding `span_err` methods.

nnethercote committed:
Rename `*note_without_error` as `*note`.

Because the variant name in `Level` is `Note`, and the `without_error`
suffix is omitted in similar cases like `struct_allow` and
`struct_help`.

nnethercote committed:
Rename `Handler::span_note_diag` as `struct_span_note`.

Because `span_note_diag` doesn't follow the naming structure used for
the error reporting functions.

nnethercote committed:
Remove an unnecessary local variable.

nnethercote committed:
Rename `HandlerInner::failure` as `HandlerInner::failure_note`.

To match the `FailureNote` variant of `Level`.

nnethercote committed:
Rename `HandlerInner::delay_span_bug` as `HandlerInner::span_delayed_bug`.

Because the corresponding `Level` is `DelayedBug` and `span_delayed_bug`
follows the pattern used everywhere else: `span_err`, `span_warning`,
etc.

nnethercote committed:
Rename `HandlerInner::delayed_span_bugs` as `HandlerInner::span_delayed_bugs`.

For reasons similar to the previous commit.

nnethercote committed:
Rename `Handler::delay_good_path_bug` as `Handler::good_path_delayed_bug`.

In line with the previous commits.

nnethercote committed:
Rename `LayoutCalculator::delay_bug` as `LayoutCalculator::delayed_bug`.

To match with the previous commits.

lnicola commented on src/tools/rust-analyzer/crates/hir-ty/src/layout.rs:
```diff
@@ -110,7 +110,7 @@ struct LayoutCx<'a> {
     type TargetDataLayoutRef = &'a TargetDataLayout;
 
-    fn delay_bug(&self, txt: String) {
+    fn delayed_bug(&self, txt: String) {
```
CC @Veykril this will be fun to handle.

nnethercote replied:
That particular commit isn't that important, so if it's a pain I can just remove it.

Veykril replied:
Isn't that only a problem when we do a sync while the changes haven't been pushed to crates io yet? fwiw we can change the autopublish time to be right before your usual sync times if you want

lnicola replied:
It is, but it's easy to patch when syncing back, then undo that when we upgrade to the published version.

nnethercote committed:
`Handler` tweaks.

- Avoid unnecessary `inner` local variables.
- Use `borrow_mut` everywhere (instead of the synonym `lock`).

nnethercote committed:
Use `Session::diagnostic` in more places.

nnethercote committed:
Use `Session::diagnostic` in more places.

nnethercote committed:
Remove unnecessary qualifiers.

nnethercote committed:
Inline and remove `LoweringContext::handler()`.

It has a single call site.

nnethercote commented:
I added a few more commits. I'll stop now, I promise :)

compiler-errors commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -996,14 +996,14 @@ impl Handler {
     }
 
-    /// For documentation on this, see `Session::delay_span_bug`.
+    /// For documentation on this, see `Session::span_delayed_bug`.
```
Can you amend this commit with a note that this used to be called `delay_span_bug`? People will probably not like that this is renamed 😓

nnethercote replied:
:+1:

nnethercote commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 61f93563d8acd5c802ce9fb5f71e1c2af28a8f26 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 61f93563d8acd5c802ce9fb5f71e1c2af28a8f26 -->
<!-- homu: {"type":"Approved","sha":"61f93563d8acd5c802ce9fb5f71e1c2af28a8f26","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 61f93563d8acd5c802ce9fb5f71e1c2af28a8f26 with merge 2da59b867621eab2c8acb2d14223ddb57cbd86e9...
<!-- homu: {"type":"BuildStarted","head_sha":"61f93563d8acd5c802ce9fb5f71e1c2af28a8f26","merge_sha":"2da59b867621eab2c8acb2d14223ddb57cbd86e9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7067585140/job/19242360993)
Approved by: compiler-errors
Pushing 2da59b867621eab2c8acb2d14223ddb57cbd86e9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7067585140/job/19242360993"},"merge_sha":"2da59b867621eab2c8acb2d14223ddb57cbd86e9"} -->

rust-timer commented:
Finished benchmarking commit (2da59b867621eab2c8acb2d14223ddb57cbd86e9): [comparison URL](https://perf.rust-lang.org/compare.html?start=bd3a22115fa3b7c572333783cd315c134b48e967&end=2da59b867621eab2c8acb2d14223ddb57cbd86e9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bd3a22115fa3b7c572333783cd315c134b48e967&end=2da59b867621eab2c8acb2d14223ddb57cbd86e9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bd3a22115fa3b7c572333783cd315c134b48e967&end=2da59b867621eab2c8acb2d14223ddb57cbd86e9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.5% | [2.5%, 2.5%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [1.3%, 2.9%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.5% | [2.5%, 2.5%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bd3a22115fa3b7c572333783cd315c134b48e967&end=2da59b867621eab2c8acb2d14223ddb57cbd86e9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=bd3a22115fa3b7c572333783cd315c134b48e967&end=2da59b867621eab2c8acb2d14223ddb57cbd86e9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 672.083s -> 671.012s (-0.16%)
**Artifact size**: 313.95 MiB -> 313.93 MiB (-0.01%)
